### PR TITLE
PERF: DataFrame transform

### LINKED ISSUE #12737: PERF: DataFrame groupby with fast transform
from [SO](http://stackoverflow.com/questions/36286803/how-to-speed-up-the-replacement-of-missing-values-for-each-groupby-group-in-pand)

```
import pandas as pd
import numpy as np

df = pd.DataFrame({'group': np.repeat(np.arange(1000), 10),
                   'B': np.nan,
                   'C': np.nan})

df.ix[4::10, 'B':'C'] = 5 # every 4th row of a group is non-null

df.groupby('group').transform('first')
```

This is then iterating over groups. Last I can see this was changed is: [here](https://github.com/pydata/pandas/pull/9994). My recollection is that this was ONLY supposed to hit in a special case, and the general case is simply a repeat based on the indices. 

This seems to be hitting in all cases makes transform back to super SLOW.

### LINKED ISSUE #13191: BUG: Series _transform_fast fails for datetime with null groups
xref #10972 - but not the same issue.  I've got a fix for this + #12737 coming.

#### Code Sample, a copy-pastable example if possible
```python
In [20]: df = pd.DataFrame({'grouping':[np.nan,1,1,3], 
                            'v':[1.1, 2.1, 3.1, 4.5], 
                            'd':pd.date_range('2014-1-1','2014-1-4')})

In [21]: df.groupby('grouping')['d'].transform('first')
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-21-cb1ed73aabc3> in <module>()
----> 1 df.groupby('grouping')['d'].transform('first')

C:\Users\Chris\Anaconda\lib\site-packages\pandas\core\groupby.pyc in transform(self, func, *args, **kwargs)
   2738                 # cythonized aggregation and merge
   2739                 return self._transform_fast(
-> 2740                     lambda: getattr(self, func)(*args, **kwargs))
   2741
   2742         # reg transform

C:\Users\Chris\Anaconda\lib\site-packages\pandas\core\groupby.pyc in _transform_fast(self, func)
   2781         out = func().values[ids]
   2782         if not mask.all():
-> 2783             out = np.where(mask, out, np.nan)
   2784
   2785         obs = np.zeros(ngroup, dtype='bool')

TypeError: invalid type promotion
```

### PR DESCRIPTION
**Pull request by chris-b1**
- [x] closes #12737, 
  closes #13191
- [ ] tests added / passed
- [x] passes `git diff upstream/master | flake8 --diff`
- [x] whatsnew entry

```
    before     after       ratio
  [2de2884 ] [4b352d9 ]
-  164.84ms     1.73ms      0.01  groupby.groupby_transform_dataframe.time_groupby_transform_dataframe
     4.44ms     3.77ms      0.85  groupby.groupby_transform_series.time_groupby_transform_series
     6.40ms     4.76ms      0.74  groupby.groupby_transform_series2.time_groupby_transform_series2
```

### CODE REVIEW

**Comment by chris-b1** [comment]
This also fixes (unless it was deliberate?) the issue of `SeriesGroupBy.transform` losing the Series's name, xref #9862

File: pandas/core/groupby.py [hunk 1]
```diff
@@ -2776,18 +2776,12 @@ def _transform_fast(self, func):
...
-        if not mask.all():
-            out = np.where(mask, out, np.nan)
-
-        obs = np.zeros(ngroup, dtype='bool')
-        obs[ids[mask]] = True
-        if not obs.all():
-            out = self._try_cast(out, self._selected_obj)
-
-        return Series(out, index=self.obj.index)
+        counts = self.size().fillna(0).values
```
Comment (jreback) [inline pandas/core/groupby.py]: not a big deal, but isn't this ``self.size().isnull().any()``?

File: pandas/core/groupby.py [hunk 2]
```diff
@@ -2776,18 +2776,12 @@ def _transform_fast(self, func):
...
+        cast = (counts == 0).any()
+        out = algos.take_1d(func().values, ids)
+        if cast:
+            out = self._try_cast(out, self.obj)
+        return Series(out, index=self.obj.index, name=self.obj.name)
```
Comment (jreback) [inline pandas/core/groupby.py]: yes, was prob missing the ``self.obj.name`` here

File: pandas/core/groupby.py [hunk 3]
```diff
@@ -3465,19 +3459,23 @@ def transform(self, func, *args, **kwargs):
...
+        # try casting data to original dtype
         counts = self.size().fillna(0).values
-        if any(counts == 0):
-            results = self._try_cast(results, obj[result.columns])
+        cast = (counts == 0).any()
+
+        # by column (could be by block?) reshape aggregated data to
+        # size of original frame by repeating obvservations with take
+        ids, _, ngroup = self.grouper.group_info
+        out = {}
```
Comment (jreback) [inline pandas/core/groupby.py]: this is going to fail with dupe columns, better to just append the arrays to a list.

**Comment by jreback** [comment]
thanks @chris-b1 very nice PR as usual!

**Pull request closed without merging**